File: frontend/src/api/db-query.ts
import request from './request'

export interface DbConnectionData {
  id?: number
  user_id?: number
  name: string
  db_type: string
  host: string
  port: number
  username: string
  password: string
  database: string
  sqlite_path: string
  created_at?: string
  updated_at?: string
}

export interface DbQueryResult {
  columns?: string[]
  rows?: any[]
  count?: number
  tables?: string[]
  table?: string
  columns_info?: { name: string; type: string; nullable: boolean; default: string }[]
  ok?: boolean
  message?: string
}

export const dbQueryApi = {
  listConnections() {
    return request.get<any, DbConnectionData[]>('/db-query/connections')
  },

  createConnection(data: DbConnectionData) {
    return request.post<any, DbConnectionData>('/db-query/connections', data)
  },

  updateConnection(id: number, data: Partial<DbConnectionData>) {
    return request.put<any, DbConnectionData>(`/db-query/connections/${id}`, data)
  },

  deleteConnection(id: number) {
    return request.delete(`/db-query/connections/${id}`)
  },

  query(action: string, config: DbConnectionData, table?: string, sql?: string) {
    return request.post<any, DbQueryResult>(`/db-query/${action}`, {
      action,
      config,
      table,
      sql,
    })
  },
}

File: backend/app/schemas/db_query.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional


class DbConnectionBase(BaseModel):
    name: str = Field(..., max_length=100)
    db_type: str = Field(..., max_length=20)
    host: str = ""
    port: int = 0
    username: str = ""
    password: str = ""
    database: str = ""
    sqlite_path: str = ""


class DbConnectionCreate(DbConnectionBase):
    pass


class DbConnectionUpdate(BaseModel):
    name: Optional[str] = None
    db_type: Optional[str] = None
    host: Optional[str] = None
    port: Optional[int] = None
    username: Optional[str] = None
    password: Optional[str] = None
    database: Optional[str] = None
    sqlite_path: Optional[str] = None


class DbConnectionResponse(BaseModel):
    id: int
    user_id: int
    name: str
    db_type: str
    host: str
    port: int
    username: str
    password: str
    database: str
    sqlite_path: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DbQueryRequestConfig(BaseModel):
    id: Optional[int] = None
    name: str = ""
    db_type: str = Field(..., max_length=20)
    host: str = ""
    port: int = 0
    username: str = ""
    password: str = ""
    database: str = ""
    sqlite_path: str = ""


class DbQueryRequest(BaseModel):
    action: str
    config: DbQueryRequestConfig
    table: Optional[str] = None
    sql: Optional[str] = None
